### ISSUE TITLE
Could not load file or assembly 'TorchSharp'

### ISSUE THREAD

**Issue by rdjaz** [dotnet/TorchSharp#1039]
Hello,

I am trying to make TorchSharp work (and by the way, thank you for this amazing framework that will help me a lot if I figure how to make it work).
After struggling, I decided to make a very simple console program and it seems that my program can't load TorchSharp assembly.

Here is the Error (It initially is in French and I translated it, but to be sure, I will put the original French error at the end):

`System.BadImageFormatException
   HResult=0x8007000B
   Message=Could not load file or assembly 'TorchSharp, Version=0.100.3.0, Culture=neutral, PublicKeyToken=67b0e5e1da516a1c' or one of its dependencies. Attempt to load an incorrectly formatted program.
   Source=<Unable to evaluate exception source>
   Procedure call tree:
<Failed to evaluate exception procedure call tree>`

I use .NET Framework 4.7.2 in Windows 10 x64 with CPU (no GPU)

Here is the code

`
using System;
using TorchSharp;
using static TorchSharp.torch.nn;
using static TorchSharp.torch.nn.functional;
using static TorchSharp.TensorExtensionMethods;
using Microsoft.DotNet.Interactive.Formatting;

namespace TorchSharpTutorial1AI
{
    internal class Program
    {
        static void Main(string[] args)
        {
            Formatter.SetPreferredMimeTypesFor(typeof(torch.Tensor), "text/plain");
            Formatter.Register<torch.Tensor>((torch.Tensor x) => x.ToString(TorchSharp.TensorStringStyle.Default, fltFormat: "G7"));

            Console.WriteLine("Starting ...");

            var tensor = torch.ones(2, 3, 3);

            Console.ReadKey();
        }
    }
}
`

And here is the installed nugget packages 👍 

![image](https://github.com/dotnet/TorchSharp/assets/123837061/a28e8e00-6421-4eb9-b5d2-f3006f647ba5)
![image](https://github.com/dotnet/TorchSharp/assets/123837061/332c528a-215d-4308-be05-d0d0cca9e411)
![image](https://github.com/dotnet/TorchSharp/assets/123837061/a0d5f116-08cb-4fa1-86d8-ccd8153a55d4)

The initial french error (in case my translation before omits something):

`System.BadImageFormatException
  HResult=0x8007000B
  Message=Impossible de charger le fichier ou l'assembly 'TorchSharp, Version=0.100.3.0, Culture=neutral, PublicKeyToken=67b0e5e1da516a1c' ou une de ses dépendances. Tentative de chargement d’un programme de format incorrect.
  Source=<Impossible d'évaluer la source de l'exception>
  Arborescence des appels de procédure :
<Impossible d'évaluer l'arborescence des appels de procédure de l'exception>`

Thank you.

**Comment by NiklasGustafsson** [comment]
When using .NET FX, you need to make sure that the application project is set to x64, not Any-CPU.

**Comment by rdjaz** [comment]
Thank you for this quick response.

Unfortunately, it's still not working:

![image](https://github.com/dotnet/TorchSharp/assets/123837061/783b53f4-4e53-41f4-b20e-19c8d1638359)
![image](https://github.com/dotnet/TorchSharp/assets/123837061/cbdfcc1e-b8d5-42e2-9ec4-b70c8b436566)

Here is the entire error message:

`System.NotSupportedException : 'This application or script uses TorchSharp but doesn't contain a reference to libtorch-cpu-win-x64, Version=2.0.1.1.

Consider referencing one of the combination packages TorchSharp-cpu, TorchSharp-cuda-linux, TorchSharp-cuda-windows or call System.Runtime.InteropServices.NativeLibrary.Load(path-to-LibTorchSharp.dll) explicitly for a Python install of pytorch. See https://github.com/dotnet/TorchSharp/issues/169.".

For CUDA, you may need to call 'TorchSharp.torch.InitializeDeviceType(TorchSharp.DeviceType.CUDA)' before any use of TorchSharp CUDA packages from scripts or notebooks.

Trace from LoadNativeBackend:

TorchSharp: LoadNativeBackend: Initialising native backend, useCudaBackend = False

Step 1 - First try regular load of native libtorch binaries.

    Trying to load native component torch_cpu relative to D:\prive\GameDev\AI\tests\TorchSharp\Tutorial1\TorchSharpTest\TorchSharpTest\bin\x64\Debug\TorchSharp.dll
    Failed to load native component torch_cpu relative to D:\prive\GameDev\AI\tests\TorchSharp\Tutorial1\TorchSharpTest\TorchSharpTest\bin\x64\Debug\TorchSharp.dll
    Trying to load native component LibTorchSharp relative to D:\prive\GameDev\AI\tests\TorchSharp\Tutorial1\TorchSharpTest\TorchSharpTest\bin\x64\Debug\TorchSharp.dll
    Failed to load native component LibTorchSharp relative to D:\prive\GameDev\AI\tests\TorchSharp\Tutorial1\TorchSharpTest\TorchSharpTest\bin\x64\Debug\TorchSharp.dll
    Result from regular native load of LibTorchSharp is False

Step 3 - Alternative load from consolidated directory of native binaries from nuget packages

    torchsharpLoc = D:\prive\GameDev\AI\tests\TorchSharp\Tutorial1\TorchSharpTest\TorchSharpTest\bin\x64\Debug
    packagesDir = D:\prive\GameDev\AI\tests\TorchSharp\Tutorial1\TorchSharpTest
    torchsharpHome = D:\prive\GameDev\AI\tests\TorchSharp\Tutorial1\TorchSharpTest\TorchSharpTest\bin
    Giving up, TorchSharp.dll does not appear to have been loaded from package directories` 

Thank you for help ; meanwhile, I still keep looking for the solution

**Comment by NiklasGustafsson** [comment]
What packages have you imported?


Sent from Outlook<http://aka.ms/weboutlook>

________________________________

**Comment by NiklasGustafsson** [comment]
It would be useful to check what DLLs are copied to the bin folder.

**Comment by NiklasGustafsson** [comment]
Can you run the examples at:

https://github.com/dotnet/TorchSharpExamples

or do they also fail?

**Comment by rdjaz** [comment]
Here's the bin folder:

![image](https://github.com/dotnet/TorchSharp/assets/123837061/641d4723-f449-4fc1-8678-e3bfd873f7f7)

And here are all the nuget packages:
![image](https://github.com/dotnet/TorchSharp/assets/123837061/de0a952d-45b1-4aed-9672-5d260fbf2704)
![image](https://github.com/dotnet/TorchSharp/assets/123837061/6cb0e4cf-f590-40d8-8d13-8cbde851835f)
![image](https://github.com/dotnet/TorchSharp/assets/123837061/37d8b409-62f4-4a3c-bbec-75c75bf5d552)

And finally, here is the references listed in the project:
![image](https://github.com/dotnet/TorchSharp/assets/123837061/8b6e8fc1-c81f-46de-8f7d-bc24141863e9)

**Comment by rdjaz** [comment]
The program that doesn't work is actually the first step of the first tutorial that I tried. The one difference is that I import the nuget package from the manager in visual studio insteed of the command prompt

**Comment by NiklasGustafsson** [comment]
Okay, so you should import either one of these combinations of NuGet packages:

1. TorchSharp + libtorch-cpu-win-x64

xor 

2. TorchSharp-cpu (this package is empty, except for its transitive dependency on what's in #1)

In the first case, your project needs to take a dependency on both, in the second case, only TorchSharp-cpu.

**Comment by rdjaz** [comment]
the 2) I think that this is what I've done. I only have imported TorchSharp-cpu. I will try a new project with the 1). If I understand I first import TorchSharp then libtorch-cpu-win-x64 only. I try and tell you.

**Comment by NiklasGustafsson** [comment]
I think the problem is that your project has taken a dependency on TorchSharp, not TorchSharp-cpu. If you want to explicitly control the backend, that is fine, but it would be simpler to just reference TorchSharp-cpu.

**Comment by NiklasGustafsson** [comment]
Maybe the explanation in this tutorial is better than what we have on the README or what I produced above:

https://github.com/dotnet/TorchSharpExamples/blob/main/tutorials/CSharp/tutorial1.ipynb

**Comment by rdjaz** [comment]
Doing the 

> Okay, so you should import either one of these combinations of NuGet packages:
> 
>     1. TorchSharp + libtorch-cpu-win-x64
> 
> 
> xor
> 
>     2. TorchSharp-cpu (this package is empty, except for its transitive dependency on what's in no. 1)
> 
> 
> In the first case, your project needs to take a dependency on both, in the second case, only TorchSharp-cpu. See the README page:
> 
> > We recommend using one of the 'bundled' packages: TorchSharp-cpu, TorchSharp-cuda-windows, or TorchSharp-cuda-linux. They will pull in the right libtorch backends.

This is what I've done but still not working. I'll try to follow the tutorial et see.

**Comment by rdjaz** [comment]
Ok. I found the problem and it's really simple. My project was in .Net Framework 4.7.2 and the packages are for .Net 6.
Do you have a version that works with 4.7 or I will have to go on .Net 6 ?

**Comment by NiklasGustafsson** [comment]
The packages are supposed to work on both .NET 6 and 4.7.2 and up (but nothing earlier on the .NET FX side).

**Comment by rdjaz** [comment]
Ok. Oddly, the simple console program works with no error on .NET 6 but not with 4.7.2 with exactly the same process. I will try to figure why it's not working on 4.7.2 a little and if it doesn't work, I will have to go on .NET6.
Thank you a lot for your help and your availability !!

**Comment by rdjaz** [comment]
On the nugget page, it tells that the .NET Framework is not targeted (if i understand well that the not targeted frameworks and versions are in light blue) :

![image](https://github.com/dotnet/TorchSharp/assets/123837061/4a55b56e-df88-4a92-b054-2b6b85e4d424)

**Comment by NiklasGustafsson** [comment]
Right, but 4.7.2 should be compliant with .NET Standard 2.0, so if we target that, it should work with 4.7.2. Same thing, we target 6.0, but it works with 7.0, too.

Regardless, we test it with 4.7.2 all the time, before each release.

**Comment by rdjaz** [comment]
Ok. So it's weird that it doesn't work on 4.7.2 on my computer when it's working on 6.0. But as I really want to quickly go into this framework to start working with it, I'll move on 6.0 even if I'm not really used to it. Maybe it's an opportunity to start working with 6.0. I don't want to loss much time on figure why it's not working on 4.7.2.
Whatever, thank you a lot for your help and for this framework that looks really great. I hope i'll have fun with it.

**Comment by NiklasGustafsson** [comment]
@rdjaz -- close?

**Comment by hoangdung3498** [comment]
> I don't know what to say -- there's nothing special about TorchSharp, it's just a collection of NuGet packages with some native code. It's being tested against 4.7.2 every build.
> 
> All you should have to do is include 'TorchSharp-cpu' or 'TorchSharp-cuda-windows' and build your solution.
 first , i have installed only TorchSharp-cpu in console app and it not work, but after migrate package.config it work. So then i tried same thing in asp.net program and i stuck again.
